src/gui/Application.cpp

```diff
@@ -388,36 +385,36 @@ bool Application::sendFileNamesToRunningInstance(const QStringList& fileNames)
     return writeOk && disconnected;
 }
 
+/**
+ * Send to-open file names to the running UI instance
+ *
+ * @param fileNames - list of file names to open
+ * @return true if all operations succeeded (connection made, data sent, connection closed)
+ */
+bool Application::sendFileNamesToRunningInstance(const QStringList& fileNames)
+{
+    return this->sendSocketCommand(SocketCmd::OpenFiles, [fileNames](QDataStream& out) { out << fileNames; });
+}
+
 /**
  * Locks all open databases in the running instance
  *
  * @return true if the "please lock" signal was sent successfully
  */
 bool Application::sendLockToInstance()
 {
-    // Make a connection to avoid SIGSEGV
-    QLocalSocket client;
-    client.connectToServer(m_socketName);
-    const bool connected = client.waitForConnected(WaitTimeoutMSec);
-    if (!connected) {
-        return false;
-    }
-
-    // Send lock signal
-    QByteArray data;
-    QDataStream out(&data, QIODevice::WriteOnly);
-    out.setVersion(QDataStream::Qt_5_0);
-    out << quint32(0); // reserve space for block size
-    out << quint32(2); // ID for database lock. TODO: move to enum
-    out.device()->seek(0);
-    out << quint32(data.size() - sizeof(quint32)); // replace the previous constant 0 with block size
+    return this->sendSocketCommand(SocketCmd::LockAll, [](QDataStream&) { /* No Data */ });
+}
 
-    // Finish gracefully
-    const bool writeOk = client.write(data) != -1 && client.waitForBytesWritten(WaitTimeoutMSec);
-    client.disconnectFromServer();
-    const bool disconnected =
-        client.state() == QLocalSocket::UnconnectedState || client.waitForConnected(WaitTimeoutMSec);
-    return writeOk && disconnected;
+/**
+ * Open and unlock a database file in the running instance
+ *
+ * @return true if the instance receives the request
+ */
+bool Application::sendUnlockToInstance(const QString& filename, const QString& password, const QString& keyfile)
+{
+    return this->sendSocketCommand(SocketCmd::Unlock,
+                                   [=](QDataStream& out) { out << filename << password << keyfile; });
```

Comment:
May need to verify this, but `[=]` also tends to captures `this` pointer and is an anti-pattern these days

Reply:
Yup I think you are good here since the auto-capture will only copy the used values which are function parameters. Using & will get you in trouble since the original values may be out of scope and de-allocated by the time they are used in the callback.

Reply:
I switched them to `&`, but is there actually a possible scoping issue with references here? My C++ is a bit a bit rusty, but there is nothing asynchronous happening with the callback.

I guess as a more general question, how does capturing behave when the variables being captured are already references? Does `=` copy the values behind the reference, or copy the reference pointer? Does `&` nest the reference to `&&var` or copy the reference directly?

Reply:
Good read: https://www.learncpp.com/cpp-tutorial/lambda-captures/

Capture by value invokes the copy constructor of non-primitive types, so a direct copy is made at the time the lambda is created. Reference takes a reference of the original value. 

Interesting tid bit in that tutorial, std::function causes a copy of the lambda to be created even if you pass by reference. This is because it implicitly wraps the lambda before passing. Didn't know that.